### PR TITLE
Use unsigned types for POINTER_TYPE

### PR DESCRIPTION
**Pull request by sbc100**
This seems to be slight codesize win because the generated code for
makeGetValue/makeSetValue I guess is slightly smaller for unsigned
types.

### CODE REVIEW

File: src/parseTools.js [hunk 1]
```diff
@@ -176,8 +176,8 @@ function needsQuoting(ident) {
 
 const POINTER_SIZE = MEMORY64 ? 8 : 4;
 const POINTER_BITS = POINTER_SIZE * 8;
-const POINTER_TYPE = 'i' + POINTER_BITS;
-
+const POINTER_TYPE = 'u' + POINTER_BITS;
+const WASM_GLOBAL_POINTER_TYPE = 'i' + POINTER_BITS;
```
Comment (kripken) [inline src/parseTools.js:180]: I don't understand what the variable name here means. Is this a pointer type for a wasm global? Or a global pointer type for wasm?

Why is it different from `POINTER_TYPE`? Maybe I don't understand the motivation here.

Comment (kripken) [reply src/parseTools.js:180]: How about this?
```js
// Similar to POINTER_TYPE, but this is the actual wasm type that is
// used in practice, while POINTER_TYPE is an internal emscripten
// type (that is unsigned, and wasm does not have unsigned types).
var POINTER_WASM_TYPE = 'i32';
```

Comment (sbc100) [reply src/parseTools.js]: Thanks!

**Review by kripken — COMMENTED** [review]
Nice, now I follow the rest of the PR! Good ideas here.

One concern though about code size. I'm surprised to see reductions here - I'd expect increases actually, since `>>> 0` for an unsigned cast is larger. So I think we should understand what causes the code size changes.

**Comment by sbc100** [comment]
> Nice, now I follow the rest of the PR! Good ideas here.
> 
> One concern though about code size. I'm surprised to see reductions here - I'd expect increases actually, since `>>> 0` for an unsigned cast is larger. So I think we should understand what causes the code size changes.

I'm failed to track down what these tiny saving are from...  the non-minified output of  hello_world_O2 seem to differ only in the names chosen for variables.  I don't see any `>>>` in the output of that test at all.  

I decided to revert those tiny codesize changes since:
1. they are in the noise
2. we accept tiny changes 
3. They are changes for the better

I'm not sure its worth diving too much deeper.. although it does show that my ability to track down the reason such changes is limited :)  Maybe I should create some better tooling for analyzing such cases.. but also I don't think we need to block this change.

**Review by kripken — APPROVED** [review]
Plan sounds good (but I still see some code size changes here?)

Yes, I was wrong about `>>> 0` - we never actually need to cast. So all you are changing here is `HEAP32` to `HEAPU32` in some cases, and I guess it's very random if that makes the distribution of names end up compressing better or not.

**Comment by dasa** [comment]
This appears to have broken `getValue(ptr, "*")` here: https://github.com/emscripten-core/emscripten/blob/main/src/library_getvalue.js#L14
It set's the `type` to "u32", and then aborts since this type isn't supported (#8640).

**Comment by sbc100** [comment]
Ah yes, I guess that line should be `POINTER_WASM_TYPE`.   We should add a test of that.

BTW, are you calling getValue() / setValue() somewhere in your own code?  How did you find this bug?

**Comment by dasa** [comment]
Yes, I'm calling `getValue` from my code and got the abort error.

**Comment by sbc100** [comment]
Thanks, fix is in #17028 

BTW, are using those functions from other JS libraries or from outside the module?

**Comment by dasa** [comment]
Thanks for the quick fix!

I'm using it from the outside. I'm using `-sEXPORTED_RUNTIME_METHODS='["getValue", "UTF8ToString"]'`

**Comment by sbc100** [comment]
Fair enough, that sounds like a valid use case.

**Comment by allsey87** [comment]
@sbc100 regarding this comment you left in the code:

> Similar to POINTER_TYPE, but this is the actual wasm type that is used in practice, while POINTER_TYPE is the more refined internal type (that is unsigned, where as core wasm does not have unsigned types).

Does this mean that pointers to memory address higher than 2GB can not be passed to/from the runtime and used to, for example, manipulate bytes via `HEAPU8`? 

This is an example the WebAssembly that I am getting out of Emscripten on version 3.1.14, I was surprised that the pointers to `invoke_iii` where not `u32` until I found this thread/your comment in the code:
```wasm
(module
  (func $env.triple (;0;) (import "env" "triple") (param i32) (result i32))
  (func $env.invoke_iii (;1;) (import "env" "invoke_iii") (param i32 i32 i32) (result i32))
  (func $env.__cxa_find_matching_catch_2 (;2;) (import "env" "__cxa_find_matching_catch_2") (result i32))
  (func $env.getTempRet0 (;3;) (import "env" "getTempRet0") (result i32))
  ;; ...
)
```

**Comment by sbc100** [comment]
The `i32` type in WebAssembly is not itself signed or unsigned, you can think of it as either being `s32` or `u32` depending on the usage, but its only only 32-bit integer type in WebAssembly, so its inevitable that you would see it in the disassembly wherever a 32-bit pointer is used.

**Pull request merged**